Add an initial support for electro-nuclear process and model

whokion opened the pull request:
This MR adds electro-nuclear cross sections using G4ElectroNuclearCrossSection, along with an initial skeleton to support the electro-nuclear process and model. The interactor will be implemented in a follow-up MR.

Commit:
Add support for the electro-nuclear process/model

Commit:
Add initial skeleton for the electro-nuclear interactor

Commit:
Add initial tests for electro-nuclear cross section calculations

Commit:
Add electro-nuclear cross section calculation

Commit:
Add electro-nuclear data, ElectroNuclearData

Commit:
Add support for the electro-nuclear process/model

Commit:
Add ElectroNuclearProcess test

Commit:
Fixed undefined reference to calc_electro_nuclear_xs

Commit:
Replace test points near the lower-bound region

Commit:
Remove version dependent test points

whokion commented:
@amandalund, @sethrj This MR is relatively large, but it closely mirrors the previous gamma-nuclear MR and consists of physics boilerplate codes. Follow-up MRs will add the details of the interactor along with additional helper codes, and onloading pipeline for Geant4 post step actions. I would also like to keep the cross-section codes as they are for now to estabilish a baseline for earlier test and validation, but plan to consolidate or simplify them as much as possible later. Thanks for your review, and sorry for another big MR.

stognini commented on src/celeritas/em/interactor/ElectroNuclearInteractor.hh:
```diff
@@ -0,0 +1,64 @@
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
```
```suggestion
// INLINE DEFINITIONS
//---------------------------------------------------------------------------//
```

whokion replied:
Yes, this code is still imcomplete.  Anyway, I added the line - also in `GammaNuclearInteractor`.

stognini commented on src/celeritas/em/model/ElectroNuclearModel.cc:
```diff
@@ -0,0 +1,157 @@
+                   << ")");
+
+    // Save particle properties
+    CELER_EXPECT(data.scalars);
```
I'd drop this, since it is doing the same thing as the validate.

whokion replied:
Remove the check - also in `GammaNuclearModel.cc`.

amandalund commented:
Thanks @whokion for this addition! Yes, the size should be mostly due to the physics boilerplate, so no worries :). I am off today but will take a look before the holidays!

whokion commented:
Thanks @amandalund ! No problem and enjoy your timeoff!

Commit:
Add missing comment separator line

Commit:
Remove redundant CELER_EXPECT check

stognini commented on src/celeritas/em/model/ElectroNuclearModel.cc:
```diff
@@ -0,0 +1,154 @@
+    // to the maximum valid energy (emax). The numbers of bins are chosen to
+    // adequately capture both parameterized points (336 bins from 2.0612 MeV
+    // to 50 GeV) and calculations used in G4ElectroNuclearCrossSection, which
+    // can be made configurable if needed (TODO).
```
Feel like this could be moved to the top documentation.

whokion replied:
Move cross-section tabulation documentation to the method description.

stognini commented on src/celeritas/em/executor/ElectroNuclearExecutor.hh:
```diff
@@ -0,0 +1,61 @@
+    // Select a target element
+    auto material = track.material().material_record();
+    auto elcomp_id = track.physics_step().element();
+    if (!elcomp_id)
```
So sometimes the element is already selected, and sometimes you need to select the element on the fly? Wouldn't it be always on the fly?

amandalund replied:
This is so in the unlikely case we need to redo the interaction (e.g. we ran out of secondary memory) we don't alter the rng stream.

stognini replied:
Ohh I see. Thanks!

sethrj replied:
This was thought to be an important issue at the beginning of the project... but what I would do now would be to have a single 2x-state-sized buffer for post-interaction secondaries+state changes, plus a "generator" that would allow things like atomic cascades.

amandalund replied:
This also shouldn't be necessary currently since we're not allocating any secondaries in the interactor, but I assume the follow-on PR will?

sethrj commented on src/celeritas/phys/PhysicsTrackView.hh:
```diff
@@ -458,11 +459,17 @@ CELER_FUNCTION real_type PhysicsTrackView::calc_xs(ParticleProcessId ppid,
+        {
+            result = MacroXsCalculator<ElectroNuclearMicroXsCalculator>(
+                params_.hardwired.electro_vd, material)(energy);
+        }
```
This and many of the other models are just tabulated nonuniform cross sections; we should be able to significantly simplify the code someday.

sethrj reviewed (commented):
Looks great @whokion !

amandalund commented on src/celeritas/em/process/ElectroNuclearProcess.hh:
```diff
@@ -0,0 +1,62 @@
+    EnergyLossGrid energy_loss(Applicability range) const final;
+
+    //! Whether the integral method can be used to sample interaction length
+    bool supports_integral_xs() const final { return false; }
```
This is false even though the incident particle is an electron or positron?

whokion replied:
This is a good question, but it would be better to set it to false. ElectroNuclearProcess is technically a hadronic inelastic process in Geant4, not a G4VEMProcess. Even though it is a energy loss process, the integral treatment may not be adequate, especially in energy regions with many resonances. To remain consistent with Geant4 and to take a conservative approach, I would set it to false.

amandalund commented on src/celeritas/em/xs/ElectroNuclearMicroXsCalculator.hh:
```diff
@@ -0,0 +1,83 @@
+  private:
+    // Shared cross section data
+    ParamsRef const& data_;
+    // Incident photon energy
```
```suggestion
    // Incident particle energy
```

whokion replied:
Fixed.

amandalund commented on src/celeritas/em/interactor/ElectroNuclearInteractor.hh:
```diff
@@ -0,0 +1,65 @@
+ */
+CELER_FUNCTION Interaction ElectroNuclearInteractor::operator()()
+{
+    return Interaction::from_onloaded();
```
Will we have a way to distinguish this type of onloading (where the incident particle continues to be tracked but the secondary is onloaded) from gamma-nuclear (where the incident particle is onloaded)?

whokion replied:
Yes, I agree. This will depend on how we handle onloaded particles, and a proper status flag should be introduced as needed in one of the follow-up MRs.

amandalund reviewed (approved):
Thanks @whokion!

Commit:
Move cross-section tabulation documentation to the method description

Commit:
Fix a typo: change photon to particle

whokion commented:
@sethrj I guess that I have addressed all review comments for now and will follow up on the remaining related suggestions in subsequent MRs. Thanks again for your review and for considering merging this at this stage.
